woocommerce/woocommerce
# [COT] Implement order search

Implements search for the COT data-store. Functionally, this should produce similar results to legacy/CPT order searches, but there are key differences worth highlighting:

Legacy (CPT) implementation:

- Search logic lives in the [data-store](https://github.com/woocommerce/woocommerce/blob/6.8.0/plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php#L581-L638).
- Initial search is composed of [two separate queries](https://github.com/woocommerce/woocommerce/blob/6.8.0/plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php#L619-L632) (order table, and order items table).
- All order and order item table rows are examined when conducting the search. So, when a search is conducted via the admin list table, the IDs need to be passed across to a final order search query.

With this implementation:

- Search logic is an integral part of [order queries](https://github.com/woocommerce/woocommerce/compare/add/33197-order-search?expand=1#diff-42f766d9a99cd9d7e77c0948a7120ddcd7123a40fe6504a46bbf715fc68ef3eaR536-R542).
- A single search query is used instead of separate queries.
- When combined with pagination, no more rows are examined than necessary (this will be especially useful within the admin list table).

See internal discussion p1657145661480799-slack-C0E1AV8T0 for context on why we took a different course. Please also note that this is not yet integrated with the order admin list table: that will happen via related issue #33197. Some other things I wanted to highlight:

- The CPT implementation looks at four meta fields whereas in this implementation we only examine two of those by default (`_billing_address_index` and `_shipping_address_index`).
- This is deliberate: the above fields have been reliably generated for many versions now and, in most cases, contain all the information we need for parity with CPT order searches.
- There is a *potential* risk that some very old legacy orders will not have those meta fields: my intention would be to follow-up with another issue to generate these either on plugin update or else as part of the COT migration.

Closes #34177.

### How to test the changes in this Pull Request:

1. Enable COT and create some test orders (unless you already have a good range of data).
2. You can programmatically conduct searches using a snippet such as `wc_order_search( 'foo' )`.
3. If the search term (`'foo'` in my example) matches part of a billing or shipping address, or the name of an order item, or else the ID of an order, then that order ID should be returned within the results.

Please also refer to the new assertions added as part of this change.

### Other information:




### FOR PR REVIEWER ONLY:
